src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,365 @@
 
 use std::cmp::Ordering;
 use std::str::from_utf8;
+use std::fs::File;
+use std::io::{Read, Seek, SeekFrom, Write};
 
 use byteorder::ByteOrder;
 use byteorder::LittleEndian;
-
 use crc::crc32;
-
 use time::Timespec;
-
 use uuid::Uuid;
 
-use consts::SECTOR_SIZE;
-
 use types::Sectors;
-use types::SectorOffset;
-
 use super::engine::DevOwnership;
+use consts::SECTOR_SIZE;
+use engine::{EngineResult, EngineError, ErrorEnum};
+use super::blockdev::BDA_STATIC_HDR_SIZE;
 
-const MDA_OFFSETS: [usize; 4] = [8, 12, 16, 20];
 const MDA_RESERVED_SIZE: Sectors = Sectors(2048 * 3); // = 3 MiB
+const NUM_MDA_REGIONS: u64 = 4;
+const MDA_REGION_HDR_SIZE: usize = 32;
 pub const MIN_MDA_SIZE: Sectors = Sectors(2040);
-const NUM_MDA_COPIES: u64 = 2;
 const STRAT_MAGIC: &'static [u8] = b"!Stra0tis\x86\xff\x02^\x41rh";
 
-#[derive(Debug, Clone, Copy)]
-pub struct MDA {
-    // Recorded values
-    pub crc: u32,
-    pub last_updated: Timespec,
-    pub used: u32,
-
-    // Computed values
-    pub offset: SectorOffset, // From start of MDA, not BDA
+#[derive(Debug, Clone)]
+pub struct BDA {
+    pub header: StaticHeader,
+    pub regions: MDARegions,
 }
 
-impl MDA {
-    pub fn new(offset: SectorOffset) -> MDA {
-        MDA {
-            crc: 0,
-            last_updated: Timespec::new(0, 0),
-            used: 0,
-            offset: offset,
-        }
-    }
-
-    fn offsets(offset: usize) -> (usize, usize, usize, usize) {
-        (offset + MDA_OFFSETS[0],
-         offset + MDA_OFFSETS[1],
-         offset + MDA_OFFSETS[2],
-         offset + MDA_OFFSETS[3])
-    }
-
-    /// Write out recorded values to buf at offset.
-    pub fn write(&self, buf: &mut [u8], offset: usize) {
-        let (offset1, offset2, offset3, offset4) = MDA::offsets(offset);
-
-        let last_updated = self.last_updated;
-        LittleEndian::write_u64(&mut buf[offset..offset1], last_updated.sec as u64);
-        LittleEndian::write_u32(&mut buf[offset1..offset2], last_updated.nsec as u32);
-
-        LittleEndian::write_u32(&mut buf[offset2..offset3], self.used);
-        LittleEndian::write_u32(&mut buf[offset3..offset4], self.crc);
+impl BDA {
```

Comment:
This could use some docs.

I am happy that it got moved out of the blockdev module.